File: moalgo1.cpp
#include <bits/stdc++.h>
using namespace std;
#define int long long
int count_[(int) 1e6 + 5];
int a[(int) 1e6 + 5];

long long removex(long long curr, long long elem) {
    curr -= count_[elem] * count_[elem] * elem;
    count_[elem]--;
    curr += count_[elem] * count_[elem] * elem;
    return curr;
}

long long addx(long long curr, long long elem) {
    curr -= count_[elem] * count_[elem] * elem;
    count_[elem]++;
    curr += count_[elem] * count_[elem] * elem;
    return curr;
}

signed main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n, q;
    cin >> n >> q;

    for (int i = 1; i <= n; i++) {
        cin >> a[i];
    }

    vector<tuple<int, int, int>> queries(q);
    for (int i = 0; i < q; i++) {
        int l, r;
        cin >> l >> r;
        queries[i] = make_tuple(l, r, i);
    }

    int block_size = sqrt(n);
    sort(queries.begin(), queries.end(), [block_size](tuple<int, int, int> a, tuple<int, int, int> b) {
        int block_a = get<0>(a) / block_size;
        int block_b = get<0>(b) / block_size;
        if (block_a != block_b)
            return block_a < block_b;
        return get<1>(a) < get<1>(b);
    });

    vector<long long> results(q);
    int current_l = 1, current_r = 0;
    long long current_strength = 0;

    for (auto [l, r, idx] : queries) {
        while (current_r < r) {
            current_r++;
            current_strength = addx(current_strength, a[current_r]);
        }
        while (current_r > r) {
            current_strength = removex(current_strength, a[current_r]);
            current_r--;
        }
        while (current_l < l) {
            current_strength = removex(current_strength, a[current_l]);
            current_l++;
        }
        while (current_l > l) {
            current_l--;
            current_strength = addx(current_strength, a[current_l]);
        }
        results[idx] = current_strength;
    }

    for (long long result : results) {
        cout << result << '\n';
    }

    return 0;
}

File: moalgo1.py
import math
import sys
input = sys.stdin.read

def mo_algorithm(n, t, a, queries):
    BLOCK_SIZE = int(math.sqrt(n))

    def add(x):
        nonlocal current_strength
        count[x] += 1
        current_strength += count[x] * count[x] * x - (count[x] - 1) * (count[x] - 1) * x

    def remove(x):
        nonlocal current_strength
        current_strength += (count[x] - 1) * (count[x] - 1) * x - count[x] * count[x] * x
        count[x] -= 1

    queries = sorted(queries, key=lambda q: (q[0] // BLOCK_SIZE, q[1]))

    currentL, currentR = 0, -1
    current_strength = 0
    count = [0] * (10**6 + 1)
    results = [0] * t

    for i, (l, r, idx) in enumerate(queries):
        while currentR < r:
            currentR += 1
            add(a[currentR])
        while currentR > r:
            remove(a[currentR])
            currentR -= 1
        while currentL < l:
            remove(a[currentL])
            currentL += 1
        while currentL > l:
            currentL -= 1
            add(a[currentL])

        results[idx] = current_strength

    return results

data = input().split()
idx = 0
n = int(data[idx])
idx += 1
t = int(data[idx])
idx += 1

a = [0] * n
for i in range(n):
    a[i] = int(data[idx])
    idx += 1

queries = []
for i in range(t):
    l = int(data[idx]) - 1
    idx += 1
    r = int(data[idx]) - 1
    idx += 1
    queries.append((l, r, i))

results = mo_algorithm(n, t, a, queries)

sys.stdout.write('\n'.join(map(str, results)) + '\n')
